### PR TITLE
tweaks to enable NativeAOT-LLVM on Linux with WASI-SDK 22

### PR DESCRIPTION
**Pull request by dicej**
Some of these changes are borrowed from https://github.com/dotnet/runtimelab/pull/2569.

Note that I had to manually copy pthread.h from the wasi-sdk-22/share/wasi-sysroot/include/wasm32-wasi-threads directory to the wasi-sdk-22/share/wasi-sysroot/include/wasm32-wasi directory as a workaround until https://github.com/WebAssembly/wasi-libc/issues/501 is addressed.

### CODE REVIEW

**Comment by dicej** [comment]
@dotnet-policy-service agree company="Fermyon Technologies"

**Comment by jkotas** [comment]
cc @dotnet/nativeaot-llvm

File: eng/native/gen-buildsys.sh
Comment (SingleAccretion) [inline eng/native/gen-buildsys.sh]: This is an upstream (dotnet/runtime) file. Why are these changes needed?

(Formulated another way: why did it work before and doesn't work now?)

File: src/coreclr/build-runtime.sh
Comment (SingleAccretion) [inline src/coreclr/build-runtime.sh]: It would be better to add `export` to the above `LLVM_CMAKE_CONFIG="$LLVM_CMAKE_CONFIG_[RELEASE|DEBUG]"` to make the script look similar to its `.cmd` version.

File: src/coreclr/jit/CMakeLists.txt
Comment (SingleAccretion) [inline src/coreclr/jit/CMakeLists.txt]: This will modify the exports for all Jits (and is cause of the current CI failure).

There are two ways to fix it:
1) Continue with the current approach of a separate `ClrJit.Llvm.exports` that is "set last".
2) Add `registerLlvmCallbacks` to the general exports file, deleting `ClrJit.Llvm.exports`, and add a stub no-op implementation of it under `#ifndef TARGET_WASM` somewhere (`ee_il_dll.cpp`).

My preference is the second approach.

File: src/coreclr/jit/llvmcodegen.cpp
Comment (SingleAccretion) [inline src/coreclr/jit/llvmcodegen.cpp]:
```suggestion
            break;
```
(Let's see if this angers MSVC's code analyzer... It can be `FALLTHROUGH;` if it does.)

We don't use LLVM macros in Jit source code.

File: src/native/libs/CMakeLists.txt
Comment (SingleAccretion) [inline src/native/libs/CMakeLists.txt]:
```suggestion
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99 -Wno-declaration-after-statement")

    // TODO-LLVM: remove once upstream moves to WASI SDK 22 as well (or otherwise fixes these warnings).
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-switch-default")
```

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:425]: This will need corresponding CI changes in `eng\pipelines\runtimelab\install-wasi-sdk.ps1` so that we test with the new SDK.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: It would be good to add a warning for older SDKs that we know will likely not work.

**Review by SingleAccretion — COMMENTED** [review]
The pthread situation will need to be resolved, s.t. an unmodified SDK can be used, by pulling the stubs from `PalRedhawkWasm.cpp` to `PalRedhawkWasm.h`.

**Comment by dicej** [comment]
Thanks for the detailed feedback, @SingleAccretion.  I'm planning to address it when I get a chance early next week.

File: eng/native/gen-buildsys.sh
Comment (dicej) [reply eng/native/gen-buildsys.sh]: The key additions are on line 100, which is a translation of the WASI-specific part of gen-buildsys.cmd.  Also, incrementally accumulating arguments like we do in `cmake_extra_defines` requires some extra care in `bash` when some of those arguments have space characters in them.  Previously, we didn't need to worry about that, but the `-DCMAKE_CROSSCOMPILING_EMULATOR=...` argument _does_ have spaces, so in order to make that work I changed `cmake_extra_defines` from a string into an array.

File: eng/native/gen-buildsys.sh
Comment (yowl) [inline eng/native/gen-buildsys.sh]:
```suggestion
        if [[ -z $WASI_SDK_PATH ]]; then
            echo "Error: Should set WASI_SDK_PATH environment variable pointing to WASI SDK root."
            exit 1
        fi
        cmake_extra_defines+=("-DCLR_CMAKE_TARGET_OS=wasi" "-DCLR_CMAKE_TARGET_ARCH=wasm" "-DWASI_SDK_PREFIX=$WASI_SDK_PATH" "-DCMAKE_TOOLCHAIN_FILE=$WASI_SDK_PATH/share/cmake/wasi-sdk.cmake" "-DCMAKE_CROSSCOMPILING_EMULATOR=node --experimental-wasm-bigint --experimental-wasi-unstable-preview1")
```

Consistent with the windows version.

**Comment by dicej** [comment]
> The pthread situation will need to be resolved, s.t. an unmodified SDK can be used, by pulling the stubs from `PalRedhawkWasm.cpp` to `PalRedhawkWasm.h`.

I tried pursuing that, but it got ugly fast.  I had to edit code under e.g. src/coreclr/gc and sprinkle new `#ifdef`s in a lot of places, which felt like overkill for a temporary workaround.

Then I took a step back and decided to add the following to `install-wasi-sdk.ps1`:

```
# Temporary WASI-SDK 22 workaround: Until
# https://github.com/WebAssembly/wasi-libc/issues/501 is addressed, we copy
# pthread.h from the wasm32-wasi-threads include directory to the wasm32-wasi
# include directory:

cp wasi-sdk/share/wasi-sysroot/include/wasm32-wasi-threads/pthread.h wasi-sdk/share/wasi-sysroot/include/wasm32-wasi/
```

If that's not acceptable, I'd vote we stick with WASI-SDK 21 for the time being, then upgrade to WASI-SDK 23, which should include a fix for https://github.com/WebAssembly/wasi-libc/issues/501.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (dicej) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: Can you elaborate on how we could check this?  Ideally we'd extract the version number from the first line of `${WASI_SDK_PATH}/VERSION` and check it against a "known-to-be-compatible" list, but I'm new to the MSBuild project file DSL, so not sure how do to that here.

File: src/coreclr/jit/CMakeLists.txt
Comment (SingleAccretion) [inline src/coreclr/jit/CMakeLists.txt]: Whitespace diff - should be reverted.

File: src/coreclr/jit/CMakeLists.txt
Comment (SingleAccretion) [inline src/coreclr/jit/CMakeLists.txt]: These warnings should be either disabled in source, addressed, or an issue filed for the removal of these suppressions.

File: eng/pipelines/runtimelab/install-wasi-sdk.ps1
Comment (SingleAccretion) [inline eng/pipelines/runtimelab/install-wasi-sdk.ps1:13]: It is ok for this hack to stay here to enable the runtime build, but:
1) There should be an issue filed tracking removing it.
2) It should be added as a step in `docs/workflow/building/coreclr/nativeaot.md`, under `download and install the Wasi SDK from...`.

**Comment by dicej** [comment]
FYI, this should address the missing pthread.h issue, once merged and included in a release: https://github.com/WebAssembly/wasi-libc/pull/504

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (SingleAccretion) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: Is `${WASI_SDK_PATH}/VERSION` a new thing - I don't see it in the v20 layout?

File: eng/native/gen-buildsys.sh
Comment (SingleAccretion) [reply eng/native/gen-buildsys.sh]: I looked at this a bit closer, and it looks like this script is only used in upstream build by `build-native.sh`, which does the same thing manually.

https://github.com/dotnet/runtime/blob/64b93517eb4bea5de51aa5bd7e5b32076198f1f9/src/native/libs/build-native.sh#L68-L79

Do we need `-DCMAKE_CROSSCOMPILING_EMULATOR=node ...`? I suspect we don't.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (dicej) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: Yeah, looks like it's new: https://github.com/WebAssembly/wasi-sdk/commit/0409839729fb4348764905a7e8c2bb53000087a3.  I guess the fallback option would be to run `${WASI_SDK_PATH}/bin/clang --version` and infer the WASI-SDK version from the LLVM version, assuming the latter maps uniquely to the former.

File: eng/native/gen-buildsys.sh
Comment (dicej) [reply eng/native/gen-buildsys.sh]: I wasn't able to get the build to work without `-DCMAKE_CROSSCOMPILING_EMULATOR=node ...`.  I believe it's used as part of CMake platform feature detection (i.e. the little test programs that CMake builds and runs to verify that the platform supports various features).

File: eng/pipelines/runtimelab/install-wasi-sdk.ps1
Comment (dicej) [reply eng/pipelines/runtimelab/install-wasi-sdk.ps1:13]: I've updated nativeaot.md and opened https://github.com/dotnet/runtimelab/issues/2598

**Comment by dicej** [comment]
Hm, I see `cl.exe` is not happy with the `#pragma clang ...` directives I added to suppress warnings in llvmlssa.cpp.  Is there a better way to approach that?

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (SingleAccretion) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: Since it looks like the version file is only present in SDK 22, we can just key off of it existing (for now).

I. e. something like:
```
<Warning Condition="!Exists('$(WASI_SDK_PATH)/VERSION')" Text="The WASI SDK version is too low. Please use WASI SDK 22 or newer with a 64 bit Clang." />
```
(I added the 64 bit breadcrumb since it looks like the 22 releases have both 32 bit and 64 bit versions.)

File: eng/native/gen-buildsys.sh
Comment (SingleAccretion) [reply eng/native/gen-buildsys.sh]: Interesting, in my local experiment, omitting this from the Windows build still succeeded. Let's leave investigating that for another day.

(Edit: after a bit more digging, I see where this is coming from.)

Changing the type of `cmake_extra_defines` has a big diff. It would be better to add a new local (`cmake_extra_defines2`...) that has the argument with spaces in it and add that to the final cmake command line.

**Comment by SingleAccretion** [comment]
>  Is there a better way to approach that?

It would be ideal if there was a way to propagate the warning upwards, but it looks like there isn't... Presumably, the directives can be guarded with `#ifdef __clang__`.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]:
```suggestion
           Condition="'$(WASI_SDK_PATH)' != '' and '$(_targetOS)' == 'wasi' and !Exists('$(WASI_SDK_PATH)/VERSION')" />
```

**Review by SingleAccretion — APPROVED** [review]
LGTM modulo one comment and assuming green CI. Thank you!

**Comment by jkotas** [comment]
Could you please resolve the conflict?

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets
Comment (yowl) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]:
```suggestion
             Condition="!Exists('$(WASI_SDK_PATH)/VERSION') and '$(_targetOS)' == 'wasi'" />
```
You probably worked this one out already.

**Comment by dicej** [comment]
I've resolved the conflict and everything's green.  Are we good to merge this?

**Comment by jkotas** [comment]
Thanks!

**Pull request merged**
